Add external schema mappings for files written with name-based schemas #40

rdsr opened the pull request:
Here's a first pass at implementing this.

External mapping is maintained as a field name -> field id map. So each alias will be an different entry in the map with the same id.  

`AssignFieldIds` assign field ids to fields. This visitor updates the schema instead of building a new schema with field ids. I felt this was OK. Let me know if this is not the case.


TODOs
1. Support schema insensitivity
2. Add tests for aliasing 
3. Maybe have a better testing strategy? . (The way I build the mapping seems a hack IMO)

In code, Open questions I've added as TODOs. I'd appreciate to if we can discuss those in the comments

rdsr commented on core/src/test/java/org/apache/iceberg/avro/TestExtNameToFieldIdsMap.java:
```diff
@@ -0,0 +1,118 @@
+    // I'm happy to change this based on comments.
+    // I'm also unsure on the direction we want to proceed with tests so
+    // for now, I've supporting the existing read projection tests
+    Map<String, Integer> namesToId = fieldNameToIds(readSchema, writeSchema);
```
This is mostly a hack, would appreciate better ways to address this.

rdsr commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -206,10 +207,22 @@ public ReadBuilder rename(String fullName, String newName) {
+     * TODO: Address case insenstivity
+     * TODO: could a schema have duplicate column names?
+     */
+    public ReadBuilder extFieldNameToId(Map<String, Integer> extFieldNameToId) {
```
The assumption I'm making is the the field names in the map are not canonical. This has the problem of not being able to store duplicate names in the schema. Let me know if we should use canonical names here.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignFieldIds.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Assign field ids to the Avro schema using the given field name to field id map
+ */
+public class AssignFieldIds extends AvroSchemaVisitor<Void> {
```
Shouldn't this return `Schema`?

rdsr replied:
I'm updating the fieldIds in place in the passed in Schema. Avro allows it. So the 'fileSchema' is being modified instead of creating a new one.

rdblue replied:
I think this should create a new one instead of modifying what was passed in.

rdsr replied:
OK, I'll update

rdblue commented on core/src/test/java/org/apache/iceberg/avro/RemoveIds.java:
```diff
@@ -0,0 +1,58 @@
+      Schema.Field field = fields.get(i);
+      resultFields.add(new Schema.Field(field.name(), fieldSchema, field.doc(), field.defaultVal()));
+    }
+    return Schema.createRecord(record.getName(), record.getDoc(), record.getNamespace(), false, resultFields);
```
I think there is a copy helper somewhere that would be useful here.

rdsr commented:
Thanks @rdblue  for the comments. I wanted to get a sense from you for the overall approach here, specifically around the mapping , where the fieldname to id map contains non canonical field names and testing strategy. Do you think the current approach is sound enough?

rdsr commented:
Hi @rdblue , please have a look when u get the time. I've addressed your remaining comments. I need your inputs specifically on some of the TODOs I've added to the code.

rdsr commented on core/src/test/java/org/apache/iceberg/avro/TestExtSchemaMapping.java:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  @Test
+  @Ignore
```
Let me know if it would be simpler to have a standalone test suite for external schema mapping instead of the current state where I extend `TestReadProjection` and ignore a couple tests

rdblue commented:
Sorry for not getting back to this very quickly. I'm planning to review it this week.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignIds.java:
```diff
@@ -0,0 +1,86 @@
+  private final Map<String, Integer> nameToId;
+  // To be used for maps and list. We do not store ids for maps and lists in our mapping, and it is not
+  // really required. For maps and list we assign a default id.
+  private static final int DEFAULT_ID = -1;
```
I think mappings need to include IDs for map keys, values and list elements. Otherwise, those can't be projected unless they contain structs with projected fields.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignIds.java:
```diff
@@ -0,0 +1,86 @@
+  public Schema map(Schema map, Schema value) {
+    Schema newMap = Schema.createMap(value);
+    newMap.addProp(AvroSchemaUtil.KEY_ID_PROP, DEFAULT_ID);
+    newMap.addProp(AvroSchemaUtil.VALUE_ID_PROP, DEFAULT_ID);
```
It isn't a good idea to assign a default ID. If a field has no mapping, the ID should be left null so that it can't be projeted.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignIds.java:
```diff
@@ -0,0 +1,86 @@
+
+  private Schema.Field newField(Schema.Field field, Schema type) {
+    Schema.Field newField = AvroSchemaUtil.copyField(field, type, field.name());
+    // TODO: Do we need to consider field.aliases()
```
Good question. I'm thinking no.

I think this should be maintained by the mapping. Mappings can have more than one name for the same field ID, and that's a better way to handle aliases than trying to keep track of all the possible combinations of aliases in the visitor and testing each one.

This assumes that when an alias is added to the Avro schema, the mapping is also updated. I think we should make that easy so that this doesn't hit users.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignIds.java:
```diff
@@ -0,0 +1,86 @@
+    Schema.Field newField = AvroSchemaUtil.copyField(field, type, field.name());
+    // TODO: Do we need to consider field.aliases()
+    Integer fieldId = nameToId.get(field.name());
+    Preconditions.checkState(fieldId != null, "No mapping found for field name %s", field.name());
```
I think that it is a problem to assume that all fields will be mapped. The upstream Avro schema can change independently from the Iceberg table schema, and we want to make sure changes like adding a field don't break readers.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -206,10 +207,22 @@ public ReadBuilder rename(String fullName, String newName) {
+     * Field name to field Id map, constructed from the table properties
+     * The current mapping is from name to field id. Aliases of a column
+     * are stored as separate entries in the map with the same value
+     * TODO: Address case insenstivity
```
I think Avro fields are case sensitive, so the mapping should be as well.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -206,10 +207,22 @@ public ReadBuilder rename(String fullName, String newName) {
+     * The current mapping is from name to field id. Aliases of a column
+     * are stored as separate entries in the map with the same value
+     * TODO: Address case insenstivity
+     * TODO: could a schema have duplicate column names?
```
No. We assume that for any name like a.b.c there is only one field. It could be ['a.b', 'c'], ['a', 'b', 'c'], ['a', 'b.c'] or ['a.b.c'], but no two of them are allowed.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -206,10 +207,22 @@ public ReadBuilder rename(String fullName, String newName) {
+     * TODO: Address case insenstivity
+     * TODO: could a schema have duplicate column names?
+     */
+    public ReadBuilder extNameToId(Map<String, Integer> extNameToId) {
```
What does "ext" stand for? External?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AssignIds.java:
```diff
@@ -0,0 +1,86 @@
+  private Schema.Field newField(Schema.Field field, Schema type) {
+    Schema.Field newField = AvroSchemaUtil.copyField(field, type, field.name());
+    // TODO: Do we need to consider field.aliases()
+    Integer fieldId = nameToId.get(field.name());
```
I think the field name should be the complete path, like `a.b.c`, not just the leaf field name.

rdblue commented on core/src/test/java/org/apache/iceberg/avro/RemoveIds.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Map;
+import org.apache.avro.Schema;
+
+class RemoveIds extends AvroSchemaVisitor<Schema> {
```
When would this be used?

rdsr replied:
This is only used in tests

rdblue commented:
@rdsr, thanks for your patience. This looks like a great start.

I made a few specific comments. In addition, I think we may want to apply the schema mapping differently. Instead of building an Avro schema with IDs, why not add an optional mapping to the conversion from Avro to Iceberg schema? Then that conversion could drop fields that don't have IDs, either embedded in the schema or from a mapping. I think that would fit a bit more cleanly because it would not be creating an intermediate file schema. It would be producing the Iceberg equivalent of the Avro schema instead. What do you think?

rdsr commented:
I think the passing an optional mapping to the conversion from Avro to Iceberg schema makes sense. Let me investigate more on this.

Thanks for the comments!

rdblue commented:
@rdsr, I've been working on #41, and I think we may have some restructuring to the mapping, instead of a simple name to ID map. Supporting aliases with a simple map requires enumerating all of the possibilities, and it is difficult to work with. I think a multi-level approach, where each struct has a mapping is going to be better to work with. I'll clean up what I have and post a PR for you to look at.

rdsr commented:
Thanks @rdblue !

rdsr commented:
@rdblue , picking this up again and going through your changes in #41 . 

I think the idea  passing an optional mapping to avro to iceberg schema is workable, and I think u mean this api  `AvroSchemaUtil#buildAvroProjection`. 

I'm working on this this week and will try to post a patch soon

rdsr commented:
@rdblue I've updated the approach here.

I now pass an optional `NameMapping`  to `BuildAvroProjection`
field ids are now determined from first the `field-id` property and then from `NameMapping` . 
The logic of how a field is selected in each of the methods (`field(), `map()`, `array()` etc) is also slightly changed. 

The general logic now is 
1. If the field does not have the field id (either as a property or in NameMapping) it is not selected 
1. if it resulting child node is null it is not selected
1. if its resulting child node is different, a new field is created.

rdblue commented:
Thanks for the update, @rdsr! I'll have a look next week when I'm back.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/Avro.java:
```diff
@@ -170,6 +171,7 @@ public static ReadBuilder read(InputFile file) {
     private final ClassLoader defaultLoader = Thread.currentThread().getContextClassLoader();
     private final InputFile file;
     private final Map<String, String> renames = Maps.newLinkedHashMap();
+    private NameMapping nameMapping;
```
Instance fields should be defaulted. Can you set this to `null`?

rdsr replied:
Is this for code readability, because AFAIK it is redundant for Java?

rdblue commented on data/src/main/java/org/apache/iceberg/data/avro/IcebergDecoder.java:
```diff
@@ -166,7 +166,7 @@ public D decode(InputStream stream, D reuse) throws IOException {
      */
     private RawDecoder(org.apache.iceberg.Schema readSchema, org.apache.avro.Schema writeSchema) {
-      this.reader = new ProjectionDatumReader<>(DataReader::create, readSchema, ImmutableMap.of());
+      this.reader = new ProjectionDatumReader<>(DataReader::create, readSchema, ImmutableMap.of(), null);
```
It would be nice to support name mapping in the message decoder as well. This is fine for now, but could you open an issue to track adding support for name mapping for Iceberg generic single-message encoding?

rdsr replied:
Added #500

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,147 @@
+            Types.NestedField.required(0, "id", Types.LongType.get()),
+            Types.NestedField.optional(5, "locations", Types.MapType.ofOptional(6, 7,
+                    Types.StructType.of(
+                            // TODO: can map keys be projected? i.e can we project only k1?
```
Yes.

rdsr replied:
Updated

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,147 @@
+  @Test
+  public void testNameMappingProjections() {
+    Schema writeSchema = new Schema(
+            Types.NestedField.required(0, "id", Types.LongType.get()),
```
Style nit: continuing indentation should be 4 spaces, not 8.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -166,17 +184,28 @@ public Schema array(Schema array, Supplier<Schema> element) {
-        Schema.Field valueProjection = element.get().getField("value");
+        Schema elementSchema = element.get();
+
+        if (fieldId(valueField) == null) {
```
What if the key field has no ID? In that case, I think it should also not be possible to project the field.

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,147 @@
+  protected GenericData.Record writeAndRead(String desc,
+                                            Schema writeSchema,
+                                            Schema readSchema,
+                                            GenericData.Record inputRecord) throws IOException {
```
I like this clever use of the existing projection tests, but it doesn't cover all cases because an ID is always available in the mapping. I think we also need to test what happens when the mapping is missing.

rdblue replied:
You should be able to use code coverage to generate some more test cases.

rdsr replied:
`testNameMappingProjections` tests this. It does not contain mapping for `locations` map

rdsr replied:
I'll also check if I can add more

rdblue commented:
@rdsr, thanks for your work on this!

I think there are some subtle problems with this because it changes assumptions about null values. For example, if an array-based map is not projected, the element struct will be an empty struct, not null. This assumes it will be null.

The main problem is that I'm having trouble reasoning about the changes in this PR to decide whether they are correct. That makes me think that combining projection with the read schema building is risky because it makes the whole thing hard to maintain.

What about instead of adding the name mapping and projection into the visitor building a read schema you instead added the mapping to PruneColumns? I think that would make sense: if a field doesn't have an ID, it is pruned; if it is mapped, then it may be pruned depending on the projected ID set. Then we wouldn't need to change the code for building a projection at all and both classes would be simpler. What do you think?

rdsr commented:
Thanks for the review @rdblue .  To be clear, I think you are referring to do something similar to my first approach where I was passing a map to `PruneColumns` to build an intermediate file avro schema, to be passed to `BuildAvroProjection`. So specifically,  this

```
Set<Integer> projectedIds = TypeUtil.getProjectedIds(expectedSchema);
// pass name mapping here build an intermediate avro file schema
Schema prunedSchema = AvroSchemaUtil.pruneColumns(newFileSchema, projectedIds, nameMapping); a
this.readSchema = AvroSchemaUtil.buildAvroProjection(prunedSchema, expectedSchema, renames);
```

But, this will require building an intermediate file **avro** schema. Is that OK ?

rdblue commented:
> I think you are referring to do something similar to my first approach where I was passing a map to PruneColumns to build an intermediate file avro schema, to be passed to BuildAvroProjection

Yes! Sorry about that, I was wrong before.

I thought that `buildAvroProjection` operated on two Iceberg schemas, so all you'd need to do is to add IDs when converting to Iceberg. But that's not correct: the primary schema is an Avro schema. (Unlike Parquet, where I think both are Iceberg and we fetch the aliases from the converted schema by ID.)

I think the snippet you posted would make this change cleaner. Sorry to cause you more work with my mistake!

rdsr commented:
No problem. Let me try to update the patch with the requested changes.

rdsr commented:
@rdblue . I updated the code and made the changes in `PruneColumns` api.

rdsr commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -100,36 +106,46 @@ public Schema union(Schema union, List<Schema> options) {
   }
 
   @Override
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
We can break this method into 2 fns if required

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -195,6 +196,8 @@ public Schema array(Schema array, Supplier<Schema> element) {
         // element was changed, create a new array
         if (elementSchema != array.getElementType()) {
+          // TODO: we do not copy field ids here. Probably not required,
+          //  but we do at other places in this class.
```
This assumes that the incoming Avro schema has field IDs. The only reason to copy in other places is when we are building a replacement schema for some reason, like if the map didn't have the LogicalMap annotation (some old manifest files), or if the value schema is a subset because only some of its fields were projected.

rdsr replied:
I've removed these todos.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -219,6 +222,7 @@ public Schema map(Schema map, Supplier<Schema> value) {
 
       // element was changed, create a new map
       if (valueSchema != map.getValueType()) {
+        // TODO: we do not copy field ids here
```
This is okay because this class assumes that the incoming schema has IDs. If you'd like, we can add Preconditions to check that assumption.

rdsr replied:
That's ok. It was just messing with my testing, making it slightly more complex. I've let comments in my testing code to highlight the problem

rdsr commented on core/src/main/java/org/apache/iceberg/avro/ProjectionDatumReader.java:
```diff
@@ -26,29 +26,33 @@
     this.fileSchema = newFileSchema;
     Set<Integer> projectedIds = TypeUtil.getProjectedIds(expectedSchema);
-    Schema prunedSchema = AvroSchemaUtil.pruneColumns(newFileSchema, projectedIds);
+    Schema prunedSchema = AvroSchemaUtil.pruneColumns(newFileSchema, projectedIds, nameMapping);
```
@rdblue do you think it would make sense to initialize `nameMapping` with `expectedSchema` if it is null. This could help us read old Avro data where table does not have an external schema mapping specified.

rdsr replied:
@rdblue any thoughts on this?

rdblue replied:
I think I mentioned it elsewhere, but I think the behavior should be:

1. If there is a name mapping in table metadata, pass it in
2. Otherwise, if the incoming schema has no field IDs, infer a mapping from the current schema and use it
3. Otherwise the schema has at least one column ID, so no inferred mapping should be used

rdsr replied:
Thanks. Is it ok if I address this as part of another commit?

rdblue replied:
Yes, of course. I don't think that needs to be part of this one. This is just adding support for a mapping if you specify one.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -165,13 +165,14 @@ public Schema array(Schema array, Supplier<Schema> element) {
+        if (keyProjection.schema() != keyField.schema() || valueProjection.schema() != valueField.schema()) {
           return AvroSchemaUtil.createProjectionMap(keyValueSchema.getFullName(),
-              AvroSchemaUtil.getFieldId(keyField), keyField.name(), keyField.schema(),
+              AvroSchemaUtil.getFieldId(keyField), keyField.name(), keyProjection.schema(),
```
The previous version didn't use `keyProjection` because the entire key must be projected. If the key is a struct of multiple columns, then projecting a subset of those columns can easily introduce key collisions that aren't in the original data.

rdsr replied:
That makes sense. I'll remove

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -58,21 +66,19 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
     }
 
-    if (hasChange) {
+    if (filteredFields.size() > 0) {
```
This shouldn't remove the `hasChange` flag. The purpose of `hasChange` is to signal when this is using the projection of a field. That way, if there are no known changed fields and all fields were projected, we use the original record. That's better for caching.

Since you now need to detect the case where the field ID is missing so that that record gets recreated, you should add a case above that detects when it isn't there and sets `hasChange = true`.

rdsr replied:
Makes sense. I've updated the patch

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -100,36 +106,46 @@ public Schema union(Schema union, List<Schema> options) {
+      Integer keyId = fieldId(keyValue.getField("key"));
+      Integer valueId = fieldId(keyValue.getField("value"));
+      if (keyId == null) {
+        Preconditions.checkState(valueId == null, "Map schema %s has value id but not key id", array);
```
This could happen if the mapping has key, but not value. I'm not sure that it's a good idea to throw an exception in this case. Maybe just warn and not project?

rdsr replied:
+1

rdsr replied:
For now I've added a log warning. Is adding logging for both [key missing and value present and value missing key present] a good idea, or do u see this to be a expected usage in which case I can remove even remove the existing logging

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -100,36 +116,48 @@ public Schema union(Schema union, List<Schema> options) {
           // the value must be a projection
           return AvroSchemaUtil.createMap(
-              keyId, keyValue.getField("key").schema(),
+              keyId, element.getField("key").schema(),
```
This should continue to use the schema from `keyValue` because map keys can't be partially projected. They must be fully projected so we always use the original key type.

rdsr replied:
Thanks, I'll update

rdsr replied:
On second thought,  we'd have to use `element.getField("key").schema(),` as the key [if it is a complex type] could contain newly assigned field ids  which the original key did not have.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -100,36 +116,48 @@ public Schema union(Schema union, List<Schema> options) {
       } else if (element != null) {
-        if (keyValue.getField("value").schema() != element.getField("value").schema()) {
+        if (keyValue.getField("key").schema() != element.getField("key").schema() ||
+            keyValue.getField("value").schema() != element.getField("value").schema()) {
```
Because the key is always fully projected, this doesn't need to check the key for changes.

rdsr replied:
key schema could be different if new field ids are assigned to fields under the key schema

rdsr replied:
Could we do this check separately? e.g make sure that key schemas are not projected (but they could contain new ids)

rdblue replied:
Yes, we will need to make sure the key schema is structurally identical. You can use `SchemaNormalization.parsingFingerprint64` for it.

rdsr replied:
Thanks I'll try this out

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -138,28 +166,98 @@ public Schema array(Schema array, Schema element) {
+
+  private Schema complexMapWithIds(Schema map, Integer keyId, Integer valueId) {
+    Schema keyValue = map.getElementType();
+    if (!AvroSchemaUtil.hasFieldId(keyValue.getField("key"))) {
```
Should this also check that the value ID is set?

rdsr replied:
I'll add, but in the org.apache.iceberg.avro.PruneColumns#map method we check that both value and key ids should be present or both should be unset. So there will not be a case where one is set and not the other

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -138,28 +166,98 @@ public Schema array(Schema array, Schema element) {
+  }
+
+  private Schema mapWithIds(Schema map, Integer keyId, Integer valueId) {
+    if (!AvroSchemaUtil.hasProperty(map, AvroSchemaUtil.KEY_ID_PROP)) {
```
Should this check that the value ID property is set?

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -138,28 +166,98 @@ public Schema array(Schema array, Schema element) {
+    }
+  }
+
+  private Integer fieldId(Schema.Field field) {
```
What about updating `AvroSchemaUtil.getFieldId` to accept a mapping? Is that used in more places where you wouldn't want to add a mapping? Seems like the changes might be simpler if the mapping was supported in util.

rdsr replied:
Today, `AvroSchemaUtil.getFieldId` is used by

1. BuildAvroProjection
2. SchemaToType
3. PruneColumns
4. SparkAvroWriter

Except 3, others will not use NameMapping. Also I will also have to pass a list of `fieldNames` together with the mapping which complicates it a little.

Let me know if you still feel that we should make this change. I'll update it then

rdsr replied:
I ended up incorporating your suggestion

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -138,28 +166,98 @@ public Schema array(Schema array, Schema element) {
-    int keyId = AvroSchemaUtil.getKeyId(map);
-    int valueId = AvroSchemaUtil.getValueId(map);
+    Integer keyId = id(map, AvroSchemaUtil.KEY_ID_PROP, "key");
+    Integer valueId = id(map, AvroSchemaUtil.VALUE_ID_PROP, "value");
```
I think the approach of adding a NameMapping lookup to the original `AvroSchemaUtil` methods would be a bit cleaner. What do you think?

rdsr replied:
I ended up taking your suggestion.

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,183 @@
+
+  //TODO: add tests on
+  // 1. aliases in namemapping
+  // 2. arrays
```
Sounds good to me.

rdsr replied:
done!

rdblue commented:
Thanks @rdsr! I think this is about ready. Just a few things to fix, like map keys in `PruneColumns`.

Sorry it took me so long to complete a review round! I kept getting pulled away mid-review.

rdsr commented:
No problem. Thanks for taking a look at it @rdblue

rdsr commented:
@rdblue,  I've addressed some of your comments. For map keys, as I have mentioned above I cannot use `keyValue` schema and a complex key could have newly assigned ids.

I've moved the 'id' determining methods to AvroSchemaUtil as u suggested.

I was also thinking whether it makes sense to initialize nameMapping with expectedSchema, if there's no nameMapping specified, under `ProjectionDatumReader` .  This could help us read old Avro data for tables which do not have external mapping specified.

rdblue commented:
> I was also thinking whether it makes sense to initialize nameMapping with expectedSchema

Yeah, I've been thinking about this suggestion as well. It would prevent getting all null values when there is no mapping (if you only have optional columns). But, it's also performing an implicit mapping that may not be correct. It's also not going to be the same behavior as we use by default for Parquet tables, but I guess we could add a table property to choose which fallback method we use for Parquet (name-based vs position-based).

In general, I think I'm for it. It would be good to hear what @aokolnychyi and @edgarRd think about this.

rdblue commented:
One more thing: for Parquet, we detect when there are no field IDs in the schema and use the fallback ID assignment strategy only in that case. If we do initialize the mapping with the projection schema, then we should do that for Avro as well. That way, we don't mix the cases where we need to fallback to name-based resolution and the case where at least some IDs are correctly assigned.

rdsr commented:
> If we do initialize the mapping with the projection schema, then we should do that for Avro as well
Do you mean Parquet ?

rdblue commented:
No, by "that" I meant checking whether there are any field IDs before using fallback. We should not use a name mapping, especially one that is created automatically, if there are already IDs in the schema.

rdsr commented:
@rdblue . Do you have any more comments for this PR?
For "initializing nameMapping with expectedSchema" we can probably address this as part of a separate PR

rdblue commented:
Sorry, I'm behind on reviews right now. I'll have a look in the next day or two.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
+      return toInt(id);
+    } else {
+      Preconditions.checkArgument(nameMapping != null,
+          "Schema does not have field id and name mapping not specified");
```
I don't think this should thrown an exception. If the name mapping isn't present or doesn't have the given field, then this should return null so that the field is pruned.

I think it should be like this:

```java
  private static Integer getId(Schema schema, String propertyName, NameMapping nameMapping, List<String> names) {
    if (schema.getType() == UNION) {
      return getId(fromOption(schema), propertyName, nameMapping, names);
    }

    Object id = schema.getObjectProp(propertyName);
    if (id != null) {
      return toInt(id);
    } else if (nameMapping != null) {
      MappedField mappedField = nameMapping.find(names);
      if (mappedField != null) {
        return mappedField.id();
      }
    }

    return null;
  }
```

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
+    return id;
+  }
+
+  static Integer getId(Schema schema, String propertyName,
```
This should be private.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
 
-  private static int getId(Schema schema, String propertyName) {
+
+  static Integer getId(Schema schema, String propertyName) {
```
I don't think there is a reason to remove the private modifier. It is still only used by other methods in this file.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
   }
 
-  private static int getId(Schema schema, String propertyName) {
+
```
Nit: extra blank line.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
+  }
+
+  static Integer getId(Schema schema, String propertyName,
+                   NameMapping nameMapping, List<String> names) {
```
Nit: this doesn't need to be wrapped and uses the wrong indentation.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java:
```diff
@@ -213,23 +239,66 @@ public static int getKeyId(Schema schema) {
+    if (id != null) {
+      return toInt(id);
+    } else {
+      Preconditions.checkArgument(nameMapping != null, "Field schema does not have id and name mapping not specified");
```
I think the same logic applies here, so this should not throw an exception. Instead, use `else if (nameMapping != null)` and return null if the object property is null or if no mapping ID is available.

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -58,10 +76,9 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
       } else if (fieldSchema != null) {
-        hasChange = true;
-        filteredFields.add(copyField(field, fieldSchema));
+        filteredFields.add(copyField(field, fieldSchema, fieldId));
```
This should still set `hasChange = true`, right?

rdsr replied:
You are right!

rdblue commented on core/src/main/java/org/apache/iceberg/avro/PruneColumns.java:
```diff
@@ -171,14 +241,19 @@ private static Schema copyRecord(Schema record, List<Schema.Field> newFields) {
       copy.addProp(prop.getKey(), prop.getValue());
     }
 
+    if (!AvroSchemaUtil.hasFieldId(field)) {
```
Should this validate that the ID is the same if it is present?

rdsr replied:
Updated

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,261 @@
+            )
+        )));
+
+    check(fileSchema, readSchema, nameMapping);
```
I think that the assertions in the `check` method are correct, but it doesn't test what data will actually be read.

This test case, for example, will project null values for all of the locations maps. Checking that the actual schema and the readSchema have the same shape doesn't validate that you get null values when you read from Avro.

I think that the solution is to rewrite writeAndRead to actually write an Avro file without the IDs and then validate that the record matches. In this case, that `locations` is null.

rdsr replied:
> This test case, for example, will project null values for all of the locations map

...

> In this case, that locations is null.

Doesn't it depend upon the file schema? The `locations` field is optional in the file schema. If the file has data for the `locations` column it will be returned, since the read/projected schema has the column projected.

rdsr replied:
Here's the read schema which gets generated when using `NameMapping`
```
{
  "type" : "record",
  "name" : "table",
  "fields" : [ {
    "name" : "id",
    "type" : "long",
    "field-id" : 0
  }, {
    "name" : "locations",
    "type" : [ "null", {
      "type" : "map",
      "values" : [ "null", {
        "type" : "record",
        "name" : "r7",
        "fields" : [ {
          "name" : "long",
          "type" : "float",
          "field-id" : 2
        } ]
      } ],
      "key-id" : 6,
      "value-id" : 7
    } ],
    "default" : null,
    "field-id" : 5
  } ]
}
```
So locations will be filled if the input record has this field.

rdblue replied:
The "locations" column should not be present. There was no field ID in the file schema and no field ID in the name mapping. That means the field cannot be projected and should be removed in column pruning.

rdsr replied:
`PruneColumn` correctly removes it. The output of `AvroSchemaUtil.pruneColumns` is this
```
{
  "type" : "record",
  "name" : "table",
  "fields" : [ {
    "name" : "id",
    "type" : "long",
    "field-id" : 0
  } ]
}
```

But the `locations` map is added back by `BuildAvroProjection` .

rdsr replied:
Since the `locations` map is part of the read schema, and if the file has a record where `locations` map is not null, even though the file schema has no id associated with the column, how can we null out the column? It seems we'd require a record level data transformation, no?

rdblue replied:
Thanks! I thought you were saying that the schema with "locations" was produced by pruning. My mistake. It looks like this is a problem with how the new pruning interacts with `BuildAvroProjection`.

`BuildAvroProjection` assumes that if a column is missing from the file schema that is passed in, it is missing from the actual file schema, but in this case it isn't. To build the projection, it adds the missing field with the expectation that the resolving Avro decoder won't read data for that column and it will be filled in with nulls. That's not going to happen because the resolving decoder will match by name.

I think all we need to do is catch this case in `BuildAvroProjection` and change the field name to something that won't match in the actual file schema. Then it will have nulls filled in automatically.

I think this is a good demonstration of why we need to test these cases with a full round-trip to a data file and back.

rdsr replied:
Hmm, let me think through this.

I think you are suggesting to change the read schema to not have the field (`locations`) if that field is not projected by NameMapping. For example, change the name to something different like `locations_[unique_suffix]` .  If that is the case, then we are effectively changing the projected/read schema. Won't this affect Spark, Presto etc, since they are expecting athe `locations` column? For instance, during query compile time the field `locations` was present as part of the read schema, but during execution or data reading it is no longer here. Do you think this is an issue?

rdsr replied:
Another point I was thinking was -- if we maintain the invariant that the read schema is always a common subset of the table schema and table mapping, then we can side step this problem.

Happy to discuss this offline if this discussion is getting too lengthy.

rdblue commented:
> If that is the case, then we are effectively changing the projected/read schema . . . Won't this affect Spark, Presto etc, since they are expecting the locations column?

Yes, that's what I'm suggesting. This won't affect Spark and Presto because the Iceberg schema of the data will still match. The name that we use to project data in Avro aren't actually used by Spark. Spark knows that id is at position 0 and location is going to be at position 1. The schema that we use to configure the Avro decoder doesn't need to be the same as the one we use for records.

Using a different schema is how we currently handle projection. Say your file has one column, 2: "a", and the current Iceberg schema is 1: "x" and 2: "y". The schema that we use to read is actually "x" and "y" (alias "a"). And that schema will be different for every file schema. The important part is that the positions are correct, not the names or concrete schema.

rdblue commented:
> if we maintain the invariant that the read schema is always a common subset of the table schema and table mapping, then we can side step this problem.

That doesn't work because the underlying records won't be large enough or won't use the right positions. The read schema needs to leave space for columns that will be filled with nulls.

rdsr commented:
Thanks for the detailed. Explanation. I'll try this out.

rdsr commented:
I've updated the `BuildAvroProjection` to rename missing optional fields.
Also updated tests to have a complete round trip to data file and back

rdblue commented on .gitignore:
```diff
@@ -10,6 +10,9 @@ out
 site/site
 
+# vscode
+.project
```
We can also consider getting these changes in separately, which would be quick.

rdsr replied:
I guess better to do it in a separate PR

rdblue commented on core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java:
```diff
@@ -115,7 +116,7 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     Types.StructType struct = current.asNestedType().asStructType();
     int fieldId = AvroSchemaUtil.getFieldId(field);
-    Types.NestedField expectedField = struct.field(fieldId); // TODO: what if there are no ids?
+    Types.NestedField expectedField = struct.field(fieldId);
```
Always good to remove these!

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,257 @@
+    // table mapping not projecting a required field 'x'
+    NameMapping nameMapping = MappingUtil.create(new Schema(
+        Types.NestedField.optional(18, "y", Types.IntegerType.get())));
+
```
Nit: extra blank line

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,257 @@
+        // Optional array field missing.
+        Types.NestedField.required(0, "id", Types.LongType.get())));
+
+    Schema readSchema = new Schema(
```
Isn't this the write schema?

rdsr replied:
Yes, in this test, read and write schema matches

rdblue replied:
Can you update this to `readSchema = writeSchema` like the test above?

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,257 @@
+            MappedField.of(22, "points", MappedFields.of(
+                MappedField.of(21, "element", MappedFields.of(
+                    // Name mapping has aliases for field x
+                    MappedField.of(19, Lists.newArrayList("x", "y", "z"))))))));
```
I don't think that y or z is required here. The ID and x are the only two pieces of information needed for the rename.

rdblue commented on core/src/test/java/org/apache/iceberg/avro/TestAvroNameMapping.java:
```diff
@@ -0,0 +1,257 @@
+    Record projected = writeAndRead(writeSchema, readSchema, record, nameMapping);
+    // field id 5 comes from read schema
+    Assert.assertNotNull("Field missing from table mapping is renamed", projected.getSchema().getField("location_r5"));
+    Assert.assertNull("location field should not be read", projected.get("location_r5"));
```
Overall, tests look good.

My remaining concern about this is that these names have changed and would break access by name like you use here. The schema has the same shape, but the requested names don't match.

Updating the schema produced by `BuildAvroProjection` is what causes this. I'm wondering if we could do slightly better, by passing in a different schema when building the [resolving decoder](https://github.com/apache/incubator-iceberg/blob/master/data/src/main/java/org/apache/iceberg/data/avro/DataReader.java#L89-L95). We can do that fairly easy for iceberg-data generic classes and the Spark support since we can update the reader builder to do these renames just before creating the resolving decoder.

Fixing Avro generics would be a bit more difficult, but not bad. Basically, we'd want to build a converter structure like the ones we use for Iceberg generics and Spark so we can use a different schema for the decoder but still create the records with the unmodified read schema. I think this would perform better than the normal path, too.

I don't think we need to do that before getting this in, but it would be nice to fix that before the next release, at least for Iceberg generics. Could you open an issue to track it and then we can get this one in without modification?

rdsr replied:
Yes, I'll create the required issues.

By Iceberg-generics u mean the apis in iceberg-data module and Avro generics is plain old Avro reader right org.apache.iceberg.avro.Avro ??

rdblue replied:
Yes.

rdblue commented:
Merging this. Nice work, @rdsr!
